### PR TITLE
Build Mixxx on Appveyor with Visual Studio 2015.

### PR DESCRIPTION
**Pull request by rryan**
Uses an x86 and x64 environment built from https://github.com/mixxxdj/buildserver/commit/3ddf42e4e06badf569b319981cae4a53559275a0 which is a combination of work by @Pegasus-RPG, @JosepMaJAZ, @sblaisot and myself. All dependencies are linked statically.

SCons is now present in the bin directory of the environment, so we don't need to install it anymore.

### CODE REVIEW

**Comment by sblaisot** [comment]
1hr time limit hit. Hummm... why is static build so much longer ?

**Comment by sblaisot** [comment]
your build env is too large. You need to shrink it by removing all strictly unneeded QT and third party libs source code.
Do you want me to takle a look at that ?

**Comment by sblaisot** [comment]
Oh, not a build env size problem.

Static linking alone takes 10 minutes for mixxx.exe and not less than 13 minutes for mixxx-test.exe ??!!

not something with can afford with appveyor. Why is static linking so long ?

**Comment by sblaisot** [comment]
Packaging failed:
`Exception : Could not find the MSVC++ runtime installer.`

You need to add vcredist.exe at the root of the build env.

File: build/appveyor/build_mixxx.bat [hunk 1]
```diff
@@ -1,61 +1,150 @@
...
 
-REM set multiprocessor to build faster (together with scons -j4)
-set CL=/MP /FS /EHsc
+REM For debugging, print the configuration log.
+type config.log
 
-scons mixxx makerelease verbose=0 winlib=%WINLIB_PATH% qtdir=%WINLIB_PATH%\build\qt-everywhere-opensource-src-4.8.6 hss1394=1 mediafoundation=1 opus=1 localecompare=1 optimize=portable build=%BUILD_TYPE% machine=%TARGET_MACHINE% toolchain=msvs virtualize=0 test=1 qt_sqlite_plugin=0 mssdk_dir=%MSSDK_DIR% build_number_in_title_bar=0 bundle_pdbs=1
+echo ==============================
+echo.
+IF ERRORLEVEL 1 (
```
Comment (sblaisot) [inline build/appveyor/build_mixxx.bat]: Here, you check the return code of echo. and not scons error code. So the tests run even if build fails

Comment (rryan) [inline build/appveyor/build_mixxx.bat]: doh, thanks!

**Comment by rryan** [comment]
> Static linking alone takes 10 minutes for mixxx.exe and not less than 13 minutes for mixxx-test.exe ??!!

It's likely link-time code-generation -- we're now doing LTCG across the entire codebase (dependencies + mixxx) whereas in a dynamic build we do LTCG for each individual DLL and mixxx separately.

File: build/appveyor/build_mixxx.bat [hunk 2]
```diff
@@ -1,61 +1,150 @@
...
+
+set OLDPATH=%PATH%
+set PATH=%BIN_DIR%;%PATH%
+scons.py mixxx makerelease toolchain=msvs winlib=%WINLIB_DIR% build=%BUILD_TYPE% staticlibs=1 staticqt=1 debug_assertions_fatal=1 verbose=1 machine=%MACHINE_TYPE% qtdir=%QTDIR% hss1394=1 mediafoundation=1 opus=1 localecompare=1 optimize=portable virtualize=0 test=1 qt_sqlite_plugin=0 mssdk_dir="%MSSDK_DIR%" build_number_in_title_bar=0 bundle_pdbs=1
+set PATH=%OLDPATH%
 
-REM set multiprocessor to build faster (together with scons -j4)
-set CL=/MP /FS /EHsc
+REM For debugging, print the configuration log.
+type config.log
```
Comment (rryan) [inline build/appveyor/build_mixxx.bat]: Yea this was just for debugging why configuration was failing. I was going to remove it, but moving it to the failure block is an even better idea

**Comment by sblaisot** [comment]
What option do we have to fit into the 1hr time limit ?
I don't see a way to achieve this atm and being unable to fit into this time limit again bells the end of windows CI :(

**Comment by rryan** [comment]
> I don't see a way to achieve this atm and being unable to fit into this time limit again bells the end of windows CI :(

Debug only builds with LTCG disabled, probably

**Comment by Be-ing** [comment]
AppVeyor build succeeded \o/

**Comment by rryan** [comment]
Ok, we now have a "fastbuild" environment built with LTCG turned off and corresponding optimize=fastbuild option. Each build completed in ~30 minutes minutes:
https://ci.appveyor.com/project/mixxxdj/mixxx/build/master-575

File: build/nsis/Mixxx.nsi [hunk 3]
```diff
@@ -119,12 +119,12 @@ Function InstallVCRedist
   SetOutPath $TEMP
 
   ; Put the VC redist installer files there
-  File ${WINLIB_PATH}\vcredist_${ARCH}.exe
+  File ${WINLIB_PATH}\vc_redist.${ARCH}.exe
```
Comment (sblaisot) [inline build/nsis/Mixxx.nsi]: please also update https://github.com/mixxxdj/mixxx/blob/master/build/wix/mixxx.wxs#L101

File: build/appveyor/install_buildenv.bat [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
 
-REM Build envs to install. You can specify more than one separated by spaces (no quotes)
-SET BUILDENVS=2.0-%1-%2-minimal
-
 REM precompiled compressed build env base URL
-SET BASEURL=https://downloads.mixxx.org/builds/appveyor/environments/2.0
+SET BASEURL=%1
+
+REM Build envs to install. You can specify more than one separated by spaces (no quotes)
+SET BUILDENVS=%2
```
Comment (sblaisot) [inline build/appveyor/install_buildenv.bat]: the above comment is not true anymore

**Comment by sblaisot** [comment]
with this build, the 64 bits package wants to install in the (32 bits) program files folder :(
I didn't already figure out why but this is currently a show stopper for merging this PR.
I'll take a look at it.

**Comment by sblaisot** [comment]
Oh, I know why : NSIS needs to be patched:
; In order for the below line to work, you must patch your C:\Program Files (x86)\NSIS\Include\MultiUser.nsh file with the one given at this link:
;   http://sourceforge.net/tracker/?func=detail&atid=373085&aid=2355677&group_id=22049

we need to include NSIS patching in appveyor setup part (or kindly ask appveyor to patch NSIS in their image)

**Comment by rryan** [comment]
> Oh, I know why : NSIS needs to be patched:
> https://github.com/rryan/mixxx/blob/628a6093cc57e8c627843e46f87166e131b20620/build/nsis/Mixxx.nsi#L34

Gotcha, so it affects all appveyor builds?

**Comment by sblaisot** [comment]
You have to revert e1b3a42 in this PR only.

File: build/appveyor/install_buildenv.bat (hunk 4 above)
Comment (rryan) [inline build/appveyor/install_buildenv.bat]: could %2 not be a space-separated list?

File: build/nsis/Mixxx.nsi (hunk 3 above)
Comment (rryan) [inline build/nsis/Mixxx.nsi]: done

**Comment by rryan** [comment]
Ok, new build environment published with both a "release" variant and a "fastbuild" variant:
http://downloads.mixxx.org/builds/appveyor/environments/2.1/

Built using [create_and_build_environments.bat](https://github.com/mixxxdj/buildserver/blob/msvc15/create_and_build_environments.bat) from this commit: https://github.com/mixxxdj/buildserver/commit/5e57821abd25e4d0b87f151ad77a6ba0911392e7 (note the short hash is in the filename of the zips).

**Comment by rryan** [comment]
Nice, build is green. I think this PR is ready to go.
https://ci.appveyor.com/project/mixxxdj/mixxx/build/master-594

**Pull request merged**